LLM/ocr_functions.py

```diff
@@ -0,0 +1,191 @@
+
+    :param image_path: Path to the image file.
+    :return: JSON object containing extracted text and table data.
+    """
+    configure_openai()
+    client = OpenAI(api_key=OPENAI_API_KEY)
+    MODEL = "gpt-4o"
+    base64_image = encode_image(image_path)
+    response = client.chat.completions.create(
+        model=MODEL,
+        messages=[
+            {"role": "user", "content": [
+                {"type": "text",
+                 "text": "Analyze this image as a completely new task. "
+                         "Identify and parse all tables and non-table data. "
+                         "For each table, carefully examine and transcribe its name from the image, typically located at the top left of the table. "
+                         "If the table name is unclear or missing, label it 'Table X' where X is a sequential number. "
+                         "Do not use any previously identified table names. "
+                         "Apply image corrections if needed for better text recognition. "
+                         "Construct each table with its newly identified name, columns, and rows, extracting all visible numbers. "
+                         "Format JSON with two main objects: 1) 'tables': an array of table objects {'table_name': '...', 'headers': [...], 'data': [[...], ...]}, 2) 'non_table_data': an object with key-value pairs for all non-table information. "
+                         "Respond only with the JSON: {'tables': [...], 'non_table_data': {...}}. "
+                         "No explanations. "
+                         "Treat this as an entirely new image with no relation to any previous tasks."
+
+                 },
+                {"type": "image_url", "image_url": {
+                    "url": f"data:image/png;base64,{base64_image}"}
+                 }
+            ]}
+        ],
+        temperature=0.0,
+        response_format={"type": "json_object"}
+    )
+    return json.loads(response.choices[0].message.content)
+
+
+def correct_image_orientation(image_path):
+    """
+    Corrects the orientation of an image based on its EXIF data.
+
+    Usage:
+    corrected_image = correct_image_orientation("path/to/image.jpg")
+
+    :param image_path: The path to the image file.
+    :return: PIL.Image.Image: The image with corrected orientation.
+    """
+    image = Image.open(image_path)
+    orientation = None
+    try:
+        for orientation in ExifTags.TAGS.keys():
+            if ExifTags.TAGS[orientation] == 'Orientation':
+                break
+        exif = dict(image.getexif().items())
+        if exif.get(orientation) == 3:
+            image = image.rotate(180, expand=True)
+        elif exif.get(orientation) == 6:
+            image = image.rotate(270, expand=True)
+        elif exif.get(orientation) == 8:
+            image = image.rotate(90, expand=True)
+    except (AttributeError, KeyError, IndexError):
+        pass
+    return image
```

Comment:
When working with any files in python, you always want to use a `with` statement or a try...close paradigm, which will close the file and prevent memory leaks. See https://realpython.com/python-with-statement/ for more details. For example: 
```suggestion
    with Image.open(image_path) as image:
      orientation = None
      try:
          for orientation in ExifTags.TAGS.keys():
              if ExifTags.TAGS[orientation] == 'Orientation':
                  break
          exif = dict(image.getexif().items())
          if exif.get(orientation) == 3:
              image = image.rotate(180, expand=True)
          elif exif.get(orientation) == 6:
              image = image.rotate(270, expand=True)
          elif exif.get(orientation) == 8:
              image = image.rotate(90, expand=True)
      except (AttributeError, KeyError, IndexError):
          pass
      return image.copy()
```

Reply:
It works in some cases. Using it directly can help eliminate errors?